package com.phucdn.microservices.order.client;

import groovy.util.logging.Slf4j;
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
import io.github.resilience4j.retry.annotation.Retry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.service.annotation.GetExchange;

@Slf4j
// @FeignClient(name = "inventory", url = "${inventory.url}")
public interface InventoryClient {
	// @RequestMapping(method = RequestMethod.GET, value = "/api/inventory")
	// boolean isInStock(@RequestParam(name = "skuCode") String skuCode,
	// @RequestParam(name = "quantity") Integer quantity);

	Logger log = LoggerFactory.getLogger(InventoryClient.class);

	@GetExchange("/api/inventory")
	@CircuitBreaker(name = "inventory", fallbackMethod = "fallbackMethod")
	@Retry(name = "inventory")
	boolean isInStock(@RequestParam(name = "skuCode") String skuCode,
			@RequestParam(name = "quantity") Integer quantity);

	default boolean fallbackMethod(String code, Integer quantity, Throwable throwable) {
		log.info("Cannot get inventory for skucode {}, failure reason: {}", code, throwable.getMessage());
		return false;
	}
}
